keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,368 @@
+---
+kep-number: 14
+title: New Resource API Proposal
+authors:
+  - "@vikaschoudhary16"
+  - "@jiayingz"
+owning-sig: sig-node
+reviewers:
+  - "@thockin"
+  - "@derekwaynecarr"
+  - "@dchen1107"
+  - "@bsalamat"
+  - "@vishh"
+approvers:
+  - "@sig-node-leads"
+editor: "@vikaschoudhary16"
+creation-date: "2018-06-14"
+last-updated: "2018-06-14"
+status: provisional
+---
+# New Resource API Proposal
+
+Table of Contents
+=================
+* [Abstract](#abstract)
+* [Background](#background)
+* [Use Stories](#user-stories)
+  * [As a cluster operator](#as-a-cluster-operator)
+  * [As a developer](#as-a-developer)
+  * [As a vendor](#as-a-vendor) 
+* [Objectives](#objectives)
+* [Non Objectives](#non-objectives)
+* [Components](#components)
+  * [ResourceClass API](#resourceclass-api)
+  * [Kubelet Extension](#kubelet-extension)
+  * [Scheduler Extension](#scheduler-extension)
+  * [Quota Extension](#quota-extension)
+* [Roadmap](#roadmap)
+
+## Abstract
+In this document we will describe a new resource API model to better support non-native compute resources on Kubernetes.
+
+## Background
+We are seeing increasing needs to better support non-native compute resources on Kubernetes that cover a wide range of resources such as GPUs, High-performance NICs, Infiniband and FPGAs. Such resources often require vendor specific setup, and have rich sets of different properties even across devices of the same type. This brings new requirements for Kubernetes to better support non-native compute resources to allow vendor specific setup, dynamic resource exporting, flexible resource configuration, and portable resource specification.
+
+The device plugin support added in Kubernetes 1.8 makes it easy for vendors to dynamically export their resources through a plugin API without changing Kubernetes core code. Taking a further step, this document proposes a new resource abstraction API, ResourceClass, that can be used to describe, manage and consume such vendor specific and metadata rich resources in simple and portable ways.
+
+## Use Stories
+### As a cluster operator:
+- Nodes in my cluster has GPU HW from different generations. I want to classify GPU nodes into one of the three categories, silver, gold and platinum depending upon the launch timeline of the GPU family eg: Kepler K20, K80, Pascal P40, P100, Volta V100. I want to charge each of the three categories differently. I want to offer my clients 3 GPU rates/classes to choose from.<br/>
+**Motivation:** As time progresses in a cluster lifecycle, new advanced, high performance, expensive variants of GPUs gets added to the cluster nodes. At the same time older variants also co-exist. There are workloads which strictly wants latest GPUs and also there are workloads which are fine with older GPUs. But since there is a wide range of types, it will be hard to manage and confusing at the same time to have granularity at each GPU type. Grouping into few broad categories will be convenient to manage.<br/>
```

Comment:
Are there any users today that need this feature from kubernetes?

Reply:
also for this user story, I think we could use `NodeAffinity` to chose different GPU type.

Reply:
This is in-line with general QoS model. We might like to experiment with this model in Openshift. /cc @derekwaynecarr 
Wondering how `NodeAffinity` can be tied with usage metrics which will be needed to charge as per usage.

Reply:
https://docs.google.com/a/google.com/document/d/1Lpuw-tjm252W4oGqBvcFAFUZkwFlpZ2i0EgpCwialN0/edit?disco=AAAAB_xFEyk

Without knowing for sure real users will benefit from it, i don't see why we'd solve this problem.

Reply:
I feel the mentioned solutions in the above comment are mostly possible workarounds for the lack of better support for flexible scheduling today. We are basically telling people today that if your workloads need any special HW resource, you need to add some special labels or taints to your nodes, add the corresponding node label selector or toleration to your workload pod spec, restrict their access through this TBD scheduler policy, make sure for a given type of resource, and each node has exactly same properties. Even with the scheduler policy proposal, cluster admins can either allow or deny pods in a namespace to specify a node selector or toleration. They still can't give different quota to different namespace. As mentioned, a main goal of this proposal is to fill the gap to support metadata rich HW resources. We want to allow them to be requested, scheduled, and quota controlled in similar ways as the existing primary resources. We want to allow cluster admins to hide the underlying HW properties from their user workloads. And we want to provide a building block on top of which people can build more flexible HW specific scheduling policies.

Reply:
Should we bring the discussion here from Google docs or keep it there?

Reply:
Perhaps having a central place of discussion is easier for other people to follow?

Reply:
See ur point, and I alway think a more general resource API model should be better than label based solutions. :)
👍

Reply:
if we are going to build a feature, we should have a clear user identified that we know will consume the feature to guide the use-case.  i think we need to evaluate the feature relative to other ideas seen in the ecosystem.

for the similar use-case of "specify gpu attributes such as gpu type and memory requirements for deployment in heterogenous GPU clusters", nvidia appears to enable this by carrying two API fields on the pod spec.

see:
https://developer.nvidia.com/kubernetes-gpu
https://github.com/NVIDIA/kubernetes/blob/875873bec8f104dd87eea1ce123e4b81ff9691d7/pkg/apis/core/types.go#L2576
https://github.com/NVIDIA/kubernetes/blob/875873bec8f104dd87eea1ce123e4b81ff9691d7/pkg/apis/core/types.go#L2389

it would be good to evaluate resource class versus this other approach.

Reply:
I'd first like to understand if gpu type and memory requirements are a real user concern today in the first place even before considering possible solutions.
There are users who are sufficiently happy with using node selectors and most users today seem to bind pods to specific gpu types either for cost or specific memory requirements.

Reply:
@derekwaynecarr we have scalability concerns with kubernetes-gpu resource. Selectors computation will be done for each compute resource in scheduler cache. OTOH with resource classes, resource classes will be fewer than compute resources.
Another concern is portability.

Reply:
Thanks a lot for the pointers, @derekwaynecarr 
As @vikaschoudhary16 mentioned, looking at nvidia's non-upstream solution, the main difference is that they changed the current resource requirement API of the container spec. In our proposal, we explicitly mentioned that this is a non-goal for the following reasons: First, in a large cluster, computing operators like “greater than”, “less than” at pod creation can be a very slow operation and is not scalable. It can cause scaling issues on the scheduler side. Second, non-primary compute resources usually lack standard resource properties. Although there are benefits to allow users to directly express their resource metadata requirements in their container spec, it may also compromise workload portability in longer term. Third, resource quota control will become harder. Fourth, we may consider the resource requirement API change as a possible extension orthogonal to the ResourceClass proposal. By introducing ResourceClass as an additional resource abstraction layer, users can express their special resource requirements through a high-level portable name, and cluster admins can configure compute resources properly on different environments to meet such requirements. We feel this helps promote portability and separation of concerns, while still maintains API compatibility.